input/connect/connect_retrode.c

```diff
@@ -0,0 +1,201 @@
+{
+   struct pad_connection* connection = (struct pad_connection*)data;
+   struct hidpad_retrode_data* device    = (struct hidpad_retrode_data*)
+      calloc(1, sizeof(struct hidpad_retrode_data));
+
+   if (!device)
+      return NULL;
+
+   if (!connection)
+   {
+      free(device);
+      return NULL;
+   }
+
+   device->connection   = connection;
+   device->slot         = slot;
+
+   port_device[port_count] = device;
+   port_count++;
+
+   return device;
+}
+
+static void hidpad_retrode_deinit(void *data)
+{
+   struct hidpad_retrode_data *device = (struct hidpad_retrode_data*)data;
+
+   if (device)
+      free(device);
+
+   port_count = 0;
+   port_device[0] = NULL;
+   port_device[1] = NULL;
+   port_device[2] = NULL;
+   port_device[3] = NULL;
+}
+
+static void hidpad_retrode_get_buttons(void *data, input_bits_t *state)
+{
+    struct hidpad_retrode_data *device = (struct hidpad_retrode_data*)data;
+    if (device)
+   {
+        BITS_COPY16_PTR(state, device->buttons);
+    }
+   else
+        BIT256_CLEAR_ALL_PTR(state);
+}
+
+static int16_t hidpad_retrode_get_axis(void *data, unsigned axis)
+{
+   int val;
+   struct hidpad_retrode_data *device = (struct hidpad_retrode_data*)data;
+
+   if (!device || axis >= 2)
+      return 0;
+
+   val = device->data[2 + axis];
+
+   // map Retrode values to a known gamepad (VID=0x0079, PID=0x0011)
+   if (val == 0x9C)
+   {
+       // axis=0 left, axis=1 up
+       val = 0x00;
+   }
+   else if (val == 0x64)
+   {
+       // axis=0 right, axis=1 down
+       val = 0xFF;
+   }
+   else
+   {
+       val = 0x7F; // no button pressed
+   }
+
+   val = (val << 8) - 0x8000;
+
+   return (abs(val) > 0x1000) ? val : 0;
+}
+
+static void hidpad_retrode_packet_handler(void *data, uint8_t *packet, uint16_t size)
+{
+   uint32_t i, pressed_keys;
+   static const uint32_t button_mapping[8] =
+   {
+           RETRO_DEVICE_ID_JOYPAD_B,
+           RETRO_DEVICE_ID_JOYPAD_Y,
+           RETRO_DEVICE_ID_JOYPAD_SELECT,
+           RETRO_DEVICE_ID_JOYPAD_START,
+           RETRO_DEVICE_ID_JOYPAD_A,
+           RETRO_DEVICE_ID_JOYPAD_X,
+           RETRO_DEVICE_ID_JOYPAD_L,
+           RETRO_DEVICE_ID_JOYPAD_R
+   };
+   struct hidpad_retrode_data *device = (struct hidpad_retrode_data*)data;
+
+   if (!device)
+      return;
+
+   // packet[1] contains Retrode port number
+   // 1 = left SNES
+   // 2 = right SNES
+   // 3 = left Genesis/MD
+   // 4 = right Genesis/MD
+
+   /*
+   // for port 1 only
+   if (packet[1] != 1)
+          return;
+
+   memcpy(device->data, packet, size);
+
+   device->buttons = 0;
+
+   pressed_keys = device->data[4];
+
+   for (i = 0; i < 8; i ++)
+      if (button_mapping[i] != NO_BTN)
+         device->buttons |= (pressed_keys & (1 << i)) ? (1 << button_mapping[i]) : 0;
+   */
+
+   // find instance which handles specific port
+   // (wiiusb_hid_read_cb calls first instance only, so need to delegate)
+   struct hidpad_retrode_data *device1234 = port_device[packet[1] - 1];
```

Comment:
C89 error, all variables must be declared at the beginning of a block.

Reply:
Fixed.